cmd/node_builder.go

```diff
@@ -77,18 +86,23 @@ type NodeBuilder interface {
 	// If the error is not nil, returns a fatal log event containing the error.
 	MustNot(err error) *zerolog.Event
 
+	// SerialStart configures the node to start components serially
+	// By default, components are started in parallel and must have explicit dependency checks.
+	// This allows using the order components were added to manage dependencies implicitly
+	SerialStart() NodeBuilder
```

Comment:
After looking at the whole solution again, I think we should remove this method.

The Builder is supposed to define each module, and specify the dependencies between them. This has to be sequential, must not be parallel, because if A depends on B, and they are both initialized in parallel, then B might not be available when initializing A.

However, running components can be done in parallel. But running the components parallel or sequential should not be determined by the component builder. The component builder should only define components and the order of creating them. Once a component manager has been built, the component builder's job is done. In other words, it's component manager's responsibility to determine whether to start each component in parallel or sequential.

I would suggest to remove this method from the builder. And make seqential start as default in component manager.

Reply:
the goal is to have components manage waiting for their dependencies explicitly, so we can get rid of the serial start altogether. To do that, we'll need the ready methods to be idempotent, and we'll need to do some work to identify the dependencies and implement the waits.

Reply:
this is now the default for nodes. I've refactored the serial start code into the node builder since that's the use case that needs it. 2e555f6